[yul-phaser] Minimal application

cameel opened the pull request:
### Description
This is the initial part of the implementation of #7806.

It provides a command-line application (`tools/yul-phaser`) which can already load `.yul` source code, apply optimizations and rank the resulting programs by their size. It runs a trivial (completely randomized) genetic algorithm and prints the result of each round to the console. ~The pull request includes a few unit tests but the coverage is still minimal.~

~This is still work in progress.~ The PR is ready.

#### Structure
- `Program` class encapsulates all the details of source code parsing, analysis and application of optimizations. An instance represents a syntax tree after a few hard-coded transformations (disambiguation, grouping, hoisting, etc.) and ready for arbitrary optimizations.
- `Chromosome` class defines a sequence of optimization steps. Provides methods for generating, changing and accessing its sequence. It does not know about `Program` and does not interact with it directly.
- `Population` class runs the genetic algorithm. It stores a set of chromosomes and refines it in subsequent algorithm rounds. It's mostly concerned with controlling the flow of the algorithm and does not itself do anything complicated. It uses `Chromosome` and `Program` to evaluate fitness of sequences.

#### Not yet implemented
1. No mutation or crossover yet. The algorithm just drops the worse half of the population in each round and refills it with randomized chromosomes.
2. No way to specify the seed for the random number generator yet.
3. ~Needs more unit tests. For now I have tests only for `Chromosome`.~
4. No saving and reloading of algorithm state. When interrupted, the state is lost.
5. A sequence is applied to the program only once. It's supposed to be applied a few times in a loop.
6. `Program` is loaded from file and parsed each time it's being evaluated. I should keep one initialized instance and make a copy when I need to apply optimizations to it.
7. Lots of stuff is still hard-coded. The program is not flexible enough for experimenting yet.

EDIT: All of the above will be implemented in subsequent PRs.

#### Known issues
1. ~`VarNameCleaner` is disabled for now because it throws `Source needs to be disambiguated` even though it is already disambiguated. `FullInliner` used to fail for me in the same way. `SSATransform` too but on `yulAssert()` instead. Both went away when I changed how I pass AST and other objects around (by value vs smart pointers/references). It's likely that I'm using it wrong or copying something that should not really be copied. I need to take a closer look at this.~

#### Questions
1. Should I worry about moving vectors of `std::string`s around? I realized that I could make almost all of them `char const*` and save some dynamic allocations because step names are just `constexpr`s but I'm not sure it's worth the time. It won't be a bottleneck anyway.
2. Is it OK to use plain `assert()` for sanity checks? `solAssert()` looks more like a normal check to me. It throws exceptions and does not get removed in release builds so I'm a bit reluctant to use it for heavier checks for invariants, preconditions and generally stuff that should not happen unless there is actually a bug in the code.
3. Since `yul-phaser` is not a library, I had to list all the necessary `.cpp` files again in `test/CMakeLists.txt` to have `soltest` link properly. This does not seem like a very good solution. Should I store those paths in some cmake variable? Or maybe try to create a library out of the classes I want to test?
4. Tests for `tools/yul_phaser/` seem like they should go into a subdirectory in `test/tools/` but that directory already exists and contains extra tools for testing so. It did not seem right to me to mix them with test cases so I put them in `tests/yul_phaser/` instead. Is that OK?

### Checklist
- [x] Code compiles correctly
- [x] All tests are passing
- [x] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [x] Used meaningful commit messages

cameel committed:
[yul-phaser] An empty command-line application for the new tool

cameel committed:
[yul-phaser] Printing help and accepting input file on the command line

cameel committed:
[yul-phaser] Add Program class

cameel committed:
[yul-phaser] Add Chromosome class

cameel committed:
[yul-phaser] Add random number generators using Mersenne Twister from boost

cameel committed:
[yul-phaser] Add Population class

cameel committed:
[yul-phaser] Create test suite for Chromosome

cameel committed:
[yul-phaser] Tests for Chromosome class

wechman commented on tools/yul_phaser/main.cpp:
```diff
@@ -15,7 +15,69 @@
+	if (arguments.count("input-file") == 0) {
+		cerr << "Missing argument: input-file." << endl;
+		return 1;
+	}
```
Może, żeby trzymać jednakowy poziom abstrakcji w cały main warto byłoby wyciągnąć helpa i parsowanie argumentów do osobnej funkcji?

axic replied:
Can we keep reviews in English please?

chriseth commented on test/CMakeLists.txt:
```diff
@@ -136,6 +136,17 @@ set(libyul_sources
 detect_stray_source_files("${libyul_sources}" "libyul/")
 
+set(yul_phaser_sources
+    yul_phaser/Chromosome.cpp
```
Can you use camelCase for directories, please?

cameel replied:
Sure. I was following the convention used in `solidity-upgrade` (#7914) but now I see that dirs elsewhere all use camel case. So this probably should be changed in `solidity-upgrade` too.

cameel replied:
Done.

chriseth commented on test/yul_phaser/Chromosome.cpp:
```diff
@@ -0,0 +1,109 @@
+
+BOOST_AUTO_TEST_CASE(test_all_steps)
+{
+	vector<string> allSteps({
```
Why don't you use the function of the Suite?

cameel replied:
Good point. I didn't want to use `Chromosome::allStepNames()` because that's a part of the code I'm testing but I somehow missed that I can just use `OptimiserSuite::allSteps()` instead of hard-coding them. Also partly because it gives me a `map` and I need a `vector` but that's not something that can't be fixed with a quick loop :)

cameel replied:
Done.

chriseth commented on tools/yul_phaser/Chromosome.h:
```diff
@@ -0,0 +1,55 @@
+	explicit Chromosome(std::vector<std::string> const& _optimizationSteps):
+		m_optimizationSteps(_optimizationSteps) {}
+	explicit Chromosome(std::vector<std::string>&& _optimizationSteps):
+		m_optimizationSteps(_optimizationSteps) {}
```
```suggestion
		m_optimizationSteps(std::move(_optimizationSteps)) {}
```

cameel replied:
Done. I have also replaced these two constructors with one that takes steps by value since that still allows vector to be moved.

chriseth commented on tools/yul_phaser/Chromosome.h:
```diff
@@ -0,0 +1,55 @@
+namespace solidity::phaser
+{
+
+class Chromosome
```
Please add short documentation

cameel replied:
Done.

chriseth commented on tools/yul_phaser/Chromosome.h:
```diff
@@ -0,0 +1,55 @@
+		m_optimizationSteps(_optimizationSteps) {}
+	explicit Chromosome(std::vector<std::string>&& _optimizationSteps):
+		m_optimizationSteps(_optimizationSteps) {}
+	static Chromosome make_random(int _length);
```
```suggestion
	static Chromosome makeRandom(int _length);
```

chriseth replied:
Also, as I assume `_length` should not be negative, please use `size_t`.

cameel replied:
Right. I was looking at `make_shared()` and `make_unique()` too much and I'm too used to snake case. I need to get used to the new convention :)

cameel replied:
> Also, as I assume `_length` should not be negative, please use `size_t`.

I'm actually a bit conflicted about this. On one hand the values I depend on internally are usually `size_t` so it makes sense to use it too. On the other hand the [guidelines say to prefer signed types for this](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#arithmetic). I eventually settled on just sticking to `size_t` but looks like I missed a few places when I was changing that. What is your stance on that? Stick to the guidelines or just don't worry and use unsigned types for non-negative values?

chriseth replied:
I actually was not aware that subscripts use signed integers. My point here, and I think this overrules the other, is that this is about the external interface of this class. If it does not make sense to call this with a negative number, then the type system should already reject it. Same rule about using pointer or reference: Use pointer only if nullptr is a valid value.

Now what the function does internally is another thing. It could cast to signed (and do length check) before doing index access, but that would probably be overkill.

So to summarize: I would use `size_t` to avoid confusion about the semantics of this function when not looking at any documentation or implementation.

cameel replied:
ok. Done.

chriseth commented on tools/yul_phaser/Chromosome.h:
```diff
@@ -0,0 +1,55 @@
+
+	size_t length() const { return m_optimizationSteps.size(); }
+
+	std::vector<std::string> const& asSequence() const { return m_optimizationSteps; }
```
```suggestion
	std::vector<std::string> const& stepNames() const { return m_optimizationSteps; }
```
or just 
```suggestion
	std::vector<std::string> const& optimizationSteps() const { return m_optimizationSteps; }
```
?

cameel replied:
I wanted to emphasize the fact that this method does not give you a copy of these steps but lets you use an existing object as if it were cast to a sequence of steps. Kinda like  `arguments["input-file"].as<string>()` in `boost::program_options`.

But well, you can look at it like an accessor too. I still prefer my name but I can change it if you think yours is clearer.

chriseth replied:
OK, but my main question here would be: It returns this as a sequence, but a sequence of what? The fact that it does not return a copy is obvious from that it returns a `const&`.

cameel replied:
Right. This should be more specific.

Anyway, after using it a bit more I now agree with your initial suggestion. I do often need to use it as an accessor and `asSequence()` isn't too great in that context. Changed to `optimisationSteps()`.

chriseth commented on tools/yul_phaser/Chromosome.h:
```diff
@@ -0,0 +1,55 @@
+	friend std::ostream& operator<<(std::ostream& _stream, Chromosome const& _chromosome);
+
+private:
+	static char abbreviateStepName(std::string const& _stepName);
```
Can you pull these out somewhere else (into the same directory where the OptimizerSuite is)? They are too generic for the Chromosome, I would say. Maybe the OptimiserSuite can even use the pulled out function then.

cameel replied:
Sure. I'll see what I can do. In that cause I should probably make `yulopti` use that too.

chriseth replied:
sounds good! yulopti could even auto-generate the help screen from the names and abbreviations at some point...

cameel replied:
Done.

I have added a pair of maps between `char` and `string` to `OptimiserSuite`. Maps are cached in static variables (just like in `allSteps()` and one is generated from the other so they won't go out of sync. I considered using `boost::bimap` or `InvertibleMap` from `libsolutil` but both seem like an overkill for a small bidirectional map that never changes.

`yulopti` has been updated, including the help screen which is now autogenerated and looks like this:
```
p: StackCompresor                q: quit                          C: ConditionalSimplifier         D: DeadCodeEliminator            
I: ForLoopConditionIntoBody      L: LoadResolver                  M: LoopInvariantCodeMotion       O: ForLoopConditionOutOfBody     
T: LiteralRematerialiser         U: ConditionalUnsimplifier       V: SSAReverser                   a: SSATransform                  
c: CommonSubexpressionEliminator d: VarDeclInitializer            e: ExpressionInliner             f: BlockFlattener                
g: FunctionGrouper               h: FunctionHoister               i: FullInliner                   j: ExpressionJoiner              
l: VarNameCleaner                m: Rematerialiser                n: ControlFlowSimplifier         o: ForLoopInitRewriter           
r: RedundantAssignEliminator     s: ExpressionSimplifier          t: StructuralSimplifier          u: UnusedPruner                  
v: EquivalentFunctionCombiner    x: ExpressionSplitter            ?
```

All the changes external to `yulPhaser` are in three commits at the bottom of the branch. I can pull them out into a separate PR if needs be.

cameel replied:
This code is now in #8215.

chriseth commented on tools/yul_phaser/Chromosome.cpp:
```diff
@@ -0,0 +1,190 @@
+	assert(_length >= 0);
+
+	vector<string> steps;
+	steps.reserve(_length);
```
Does this save so much?

cameel replied:
Probably nothing measurable. 4 or 5 reallocations since there are 28 possible steps right now. It's just that not pre-allocating the memory when I know the exact size ahread of time seems kinda wrong to me :)

Should I drop it?

chriseth replied:
Once we determine that this function is crucial for performance, we can reevaluate, but at this point, this is more noise than helpful. It might even be that the compiler/optiimzer can determine the length from looking at the loop...

cameel replied:
Removed this and a few other similar calls to `reserve()`. From now on I'll use it only if it's likely to affect performance.

chriseth commented on tools/yul_phaser/Chromosome.cpp:
```diff
@@ -0,0 +1,190 @@
+
+	vector<string> steps;
+	steps.reserve(_length);
+	for (auto i = 0; i < _length; ++i)
```
```suggestion
	for (size_t i = 0; i < _length; ++i)
```

cameel replied:
Done.

chriseth commented on tools/yul_phaser/Chromosome.cpp:
```diff
@@ -0,0 +1,190 @@
+
+vector<string> Chromosome::allStepNames()
+{
+	assert(not OptimiserSuite::allSteps().empty());
```
`not`? :)

cameel replied:
You mean why I didn't use `!`? I think that `not`, `and` and `or` read better than `!`, `&&` and `||`. But that's just my personal preference. I can switch if you prefer to just have `!`, `&&`, `||` consistently everywhere.

BTW this assertion is no longer necessary. I changed the code here and forgot to remove it. Originally `stepNames` below was a `static` variable initialized the first time the function runs. If it's empty, it means that it has not been initialized yet. But that's not necessarily true if `OptimiserSuite::allSteps()` for whatever reason happens to be empty too. That probably won't happen under normal circumstances and even it did, it won't break the code, just make it unnecessarily reinitialize the variable but I wanted to make this corner case clear.

cameel replied:
Looks like MSVC 2017 in AppVeyor does not like it :(
Replaced with `!`, `&&` and `||`.

chriseth commented on tools/yul_phaser/Chromosome.cpp:
```diff
@@ -0,0 +1,190 @@
+	};
+
+	auto pair = stepToSymbol.find(_stepName);
+	assert(pair != stepToSymbol.end());
```
If you use `assert`, you could also use `.at`.

cameel replied:
Done.

chriseth commented on tools/yul_phaser/Chromosome.cpp:
```diff
@@ -0,0 +1,190 @@
+
+vector<string> Chromosome::allStepNamesExcept(vector<string> const& _excludedStepNames)
+{
+	vector<string> all = allStepNames();
```
Maybe - half jokingly:
```
return convert_container<vector<string>>(convert_container<set<string>>(allStepNames()) - convert_container<set<string>>(_excludedStepNames));
```

cameel replied:
I wish! Using set arithmetic is actually the first thing that comes to my mind in a situation like this but looks like `std::set` does not implement the `-` operator :( There's `set_difference()` but it's not as concise so I settled for a boring `for` :)

chriseth replied:
We have an overload for `operator-` in one of the util headers.

cameel replied:
Done.

Actually I removed this function a commit later anyway because after removing `VarNameCleaner` nothing needs to be excluded any more, but it's still in commit history in case I need it back. I bet it'll come in handy for debugging later.

chriseth commented on tools/yul_phaser/Program.h:
```diff
@@ -0,0 +1,95 @@
+
+	std::shared_ptr<yul::Block> m_ast;
+	yul::NameDispenser m_nameDispenser;
+	yul::OptimiserStepContext m_optimiserStepContext;
```
I think the source of the bug is here: Program is copyable but m_optimiserStepContext contains a reference to m_nameDispenser.

cameel replied:
Thanks for spotting this. Yeah, the way this class handled copy construction was broken. I fixed this by recreating the context in the copy constructor (instead of making a copy) but this still leaves me with shared `m_ast`. I should clone it but I don't yet know how to do it. I could just walk the tree and do it manually but I bet someone has already implemented it somewhere. For now I just made the class `noncopyable` and I'll get back to this problem later.

This was not the cause of `VarNameCleaner` crashes though. I spent a while debugging it and reading through `SSATransform` which fails on `YulAssertion`. It fails because `NameDispenser` gives it names already present in the code. And that's `VarNameCleaner`'s fault. This step changes names in the code but does not mark them as used in the dispenser.

This is completely reproducible in `yulopti` too. Just run
``` bash
build/test/tools/yulopti test/libyul/yulOptimizerTests/ssaTransform/for_reassign_init.yul
```
and type `hgoalaVaalVVVa`. It will crash.

I guess it was not noticed (or maybe just not considered a problem) before because AFAIK it's just final cleanup that does not affect the structure of the AST (just makes the names cleaner) so it's not likely to be anything but the last step.

How should it be fixed?
1) Make `VarNameCleaner` add newly created names to the dispenser. This leaves the removed names in. Not sure if this is a problem or not.
2) Make `VarNameCleaner` both add and remove names to and from the dispenser. This requires a change in dispenser's API which does not have any methods for removing stuff.
3) Just make sure never to run any other optimisation steps after `VarNameCleaner`. I already made it a special case in the `Chromosome`. But that's risky.

chriseth replied:
The VarNameCleaner should not be considered by this tool - it is not a real optimization step, just as the disambiguator is not. Actually, it does not even guarantee unique names any more, so you would have to run the disambiguator after it.

cameel replied:
Done and moved to a separate PR: #8214.

chriseth commented:
Have you thought about using the string of abbreviated step names for the chromosome instead of the vector of strings of full names?

cameel committed:
[yul-phaser] Program: Use unique_ptr rather than shared_ptr to hold a pointer to the AST root

- The class never shares the instance so unique_ptr makes more sense.

cameel committed:
[yul-phaser] Make Program noncopyable

- Copying worked but resulted in OptimiserStepContext having a reference to NameDispenser instance of the other object.

cameel committed:
[yul-phaser] Make Program and Population classes accept source code rather than file path

- I need some sample .yul files for testing but I see that existing tests generally have source code hard-coded in them rather than in standalone .yul files. There are lots of .yul files but they seem to be automatically processed by a special test case rather loaded ad-hoc by manually created tests.
- Program and Population required a file name until now. I'm making them accept loaded source code to be able to give them data hard-coded in a test.

cameel committed:
[yul-phaser] Printing and JSON conversion for the Program class

cameel committed:
[yul-phaser] Tests for Program class

cameel committed:
[yul-phaser] Tests for random number generators

cameel committed:
[yul-phaser] Tests for Population class

cameel committed:
[yul-phaser] Create OptimiserStepContext on demand instead of storing it in Program class

- This also lets us get rid of the static s_externallyUsedIdentifiers.

cameel commented:
> Have you thought about using the string of abbreviated step names for the chromosome instead of the vector of strings of full names?

I thought about storing pointers to strings or `std::reference_wrapper`s (since they strings are static). But a vector of strings looked like it was meant to be the a canonical way to refer to optimisation steps and I didn't want to unnecessarily invent anything new here so I went with it. Now that you mention it, storing chars would be more efficient and still elegant enough with regards to the class API. I'll consider changing it but not in this PR.

cameel commented:
I just pushed all my changes for this PR. It's now ready for another round of review. It adds a full set of tests, fixes all current review issues and adds a lot of small improvements.

All the modifications are in fixups and new commits so in theory you only need to look at them but there are a lot of fixups so it might be better to just review the whole thing again.

Here's a summary of changes not already mentioned in review comments above to make it easier to notice the differences:
1. New stuff
    - Tests for `Population`, `Program`, `Random` and `Chromosome`.
        - Some tests that deal with random stuff are probabilistic, but they're constructed in such a way that a false positive is not likely to happen within this lifetime of the universe :)
    - Added docstrings for classes.
2. Changes outside of `yul-phaser`
    - Changed `yul::Parser::parse()` factory method to return `unique_ptr` rather than `shared_ptr`. This way the caller can decide which one to use instead of being forced to use `shared_ptr`.
3. Changes to make stuff more testable
    - `Population.run()` takes a stream parameter instead of having `cout` hard-coded.
    - `ast()`, `operator<<` and `toJson()` in `Program` class.
    - `operator==` and `!=` in `Chromosome` class.
    - `individuals()`, `operator<<` in `Population` class. `measureFitness()` made `public`.
4. Small improvements and optimisations
    - A few more forward declarations instead of including header.
    - Always use references when declaring loop variables with `auto`.
5. Refactoring
    - Refactored `Population` constructors.
    - `protected` stuff in `Population` is now `private`.
    - Separate function for argument parsing in `main.cpp`.
    - `Program` is not supposed share its pointers so it now uses `unique_ptr` for everything. This required the change to `yul::Parser::parse()` mentioned above.
    - Removed the redundant `test_` prefix from my test case names. Wrapped all test suites in another one called `Phaser` to make it possible to run all phaser tests with `--run_test=Phaser`.
    - `BOOST_TEST` instead of `BOOST_CHECK_EQUAL` in tests.
    - `Population::MeanChromosomeLength` should have been called `Population::MaxChromosomeLength` since it's used as the number of trials in binomial distribution.

axic commented on test/tools/yulopti.cpp:
```diff
@@ -139,112 +175,37 @@ class YulOpti
+			map<char, string> const& abbreviationMap = OptimiserSuite::stepAbbreviationToNameMap();
+			map<char, string> const& extraOptions = {
+				{'q', "quit"},
+				{'p', "StackCompresor"},
```
```suggestion
				{'p', "StackCompressor"},
```

cameel replied:
Thanks for catching this. Fixed.

chriseth commented on tools/yulPhaser/Chromosome.cpp:
```diff
@@ -0,0 +1,148 @@
+#include <tools/yulPhaser/Random.h>
+
+#include <libyul/optimiser/BlockFlattener.h>
+#include <libyul/optimiser/CommonSubexpressionEliminator.h>
```
Do we need all these includes?

cameel replied:
Right, we don't. I only needed them for debugging. I'll remove them.

cameel replied:
Done.

chriseth commented:
It might be a good idea to extract the changes to yulopti into its own PR. Also could you rebase and squash the rest?

chriseth commented on tools/yulPhaser/main.cpp:
```diff
@@ -0,0 +1,130 @@
+	po::variables_map arguments;
+};
+
+CharStream loadSource(string const& _sourcePath)
```
Maybe put inside anonymous namespace?

cameel replied:
Sure. I did it for other `.cpp` files but it slipped my mind that `main.cpp` is does not get any special treatment from linker in that regard.

cameel replied:
Done.

chriseth commented on libyul/optimiser/Suite.cpp:
```diff
@@ -372,6 +374,72 @@ map<string, unique_ptr<OptimiserStep>> const& OptimiserSuite::allSteps()
+///
+/// @pre @a originalMap must have unique values.
+template <typename K, typename V>
+map<V, K> invertMap(map<K, V> const& originalMap)
```
Can you put that into CommonData.h?

cameel replied:
Sure. I considered it but I wasn't sure if I should. It's generic but that need might not be all that common (especially given there's `InvertibleMap` already).

cameel replied:
Done and moved to a separate PR together with the step abbreviations: #8215.

chriseth commented on libyul/optimiser/Suite.cpp:
```diff
@@ -372,6 +374,72 @@ map<string, unique_ptr<OptimiserStep>> const& OptimiserSuite::allSteps()
+map<V, K> invertMap(map<K, V> const& originalMap)
+{
+	map<V, K> inverseMap;
+	for (auto const& originalPair: originalMap)
```
Maybe 
```suggestion
	for (auto const& [key, value]: originalMap)
```

cameel replied:
Done in #8215.

chriseth commented on tools/yulPhaser/Chromosome.cpp:
```diff
@@ -0,0 +1,148 @@
+		//{UnusedPruner::name},
+		//{VarDeclInitializer::name},
+
+		// VarNameCleaner is not safe to use here because it leaves name dispenser in an unusable
```
We might want to take this step out of this tool altogether, it is just a prettifier.

cameel replied:
Should I remove it from `OptimizerSuite` in this PR?

chriseth replied:
Yes, please!

chriseth commented on tools/yulPhaser/Population.cpp:
```diff
@@ -0,0 +1,140 @@
+using namespace std;
+using namespace solidity::langutil;
+
+namespace solidity::phaser
```
```suggestion
using namespace solidity::phaser;
```

cameel replied:
Right. I didn't notice you were using this style. I'll switch to it in my code.

cameel replied:
Done. There's one downside to this style though. It forces me to add forward declarations of all `friend`s which usually requires forward declaring the class too.

```
struct Individual;
class Population;
std::ostream& operator<<(std::ostream& _stream, Individual const& _individual);
std::ostream& operator<<(std::ostream& _stream, Population const& _population);
```

I don't need to do this if the definition is wrapped in a namespace too.

I wondered why this did not came up in the rest of the codebase. Looks like you mostly just define `operator<<` inline. I didn't want to pollute the header with its implementation details though.

chriseth replied:
Hm, I don't see why a change in the cpp file demands a change in the header. This probably has not come up because we don't really use `operator<<`.

cameel replied:
I just checked and the forward declaration does not really need to be in the header. Looks like I could have put it in the `.cpp` file just as well. The header was a natural choice for a declaration but I just realized that putting it in the `.cpp` file would have saved me the trouble of having to forward-declare the class. I'll move the declarations there. Still, a declaration inside the namespace is required.

It has not come up so far because the existing code always defines friends inline inside the namespace (inside or outside of class body but still in the header file). And a definition outside of the namespace requires a declaration in that namespace even if you do `using namespace`:

``` c++
namespace NX { void foo(); }
namespace NY {}
namespace NZ { class A { friend void baz(); }; }

using namespace NX;
using namespace NY;
using namespace NZ;

void NX::foo() {}
void NY::bar() {} // ERROR: 'void NY::bar()' should have been declared inside 'NY'
void NZ::baz() {} // WARNING: 'void NZ::baz()' has not been declared within 'NZ'
```
The important detail is that the `friend` declaration in a class does not count as a declaration in the namespace. A separate declaration is still needed if you want to define the function outside of the namespace as in my case.

chriseth commented on tools/yulPhaser/Population.cpp:
```diff
@@ -0,0 +1,140 @@
+
+size_t Population::measureFitness(Chromosome const& _chromosome, CharStream& _sourceCode)
+{
+	auto program = Program::load(_sourceCode);
```
Can the result of `Program::load` be stored across multiple chromosome evaluations? Is that what you are currently working on with the ASTCopier?

cameel replied:
Yes, that's the idea. I'll have one "master copy" obtained from `load()` and then `measureFitness()` will copy it and only run `optimize()` on it.

It will come in the next PR though because this change will affect several files. I could make it a small PR immediately after this one though if you want. Otherwise it'll be included in the PR with the rest of the algorithm changes.

chriseth replied:
Ok!

cameel replied:
The PR is here: #8223.

chriseth commented on tools/yulPhaser/Population.cpp:
```diff
@@ -0,0 +1,140 @@
+namespace solidity::phaser
+{
+
+size_t Population::measureFitness(Chromosome const& _chromosome, CharStream& _sourceCode)
```
I don't see much sense in having this static if you pass almost half the state anyway :)

cameel replied:
Yeah, this is not the best design but in this PR the `Population` class is still just a stopgap solution to have the algoritm do something.

The mutation/crossover/evaluation part will change completely. Fitness metric will be a separate base class with various specializations (e.g. code size). It will be passed to population's constructor. Similarly, I'll have base classes for mutations, crossover and selections and specializations. This will allow me to use them as building blocks to easily construct and test different variations of the algorithm in later phases (in this phase I'm just doing the base classes and one or two specializations for each). I'm also going to make evaluation implicit (store chromosomes in an ordered set, always evaluated and sorted and make population evaluate new chromosomes as soon as they are created/added/mutated).

Below is an example of how an algorithm will look like using this approach.

``` c++
// One round of the algorithm, inside GeneticAlgorithm::run()

Population populationWithMutations = population;
populationWithMutations.mutateInPlace(RandomSelection(), GeneRandomization(RandomGenes(0.1)));
populationWithMutations.mutateInPlace(RandomSelection(), GeneDeletion(RandomGenes(0.1)));
Population recombinations = populationWithMutations.crossover(RandomPairSelection(), HalfSwap(0.5));

return (populationWithMutations + recombinations).select(TopSelection());
```

This would be initialized with:
``` c++
auto sourceCode = make_shared<Program>(program);
GeneticAlgorithm algorithm(ProgramSize(sourceCode), cout);
algorithm.run();
```

But that's a big change in the upcoming PR. In this one I'd prefer to just leave it as is.

chriseth commented on tools/yulPhaser/Population.cpp:
```diff
@@ -0,0 +1,140 @@
+	doEvaluation();
+	for (size_t round = 0; !_numRounds.has_value() || round < _numRounds.value(); ++round)
+	{
+		doMutation();
```
Shouldn't it be mutation - evaluation - selection?

cameel replied:
Yes, but I wanted to see the fitness for all chromosomes in the output and didn't really want to run evaluate twice :) So I put it after selection but this forced me to do one evaluation call before the loop.

It's not great but, but this code will change completely in the next PR anyway so I'd leave it as is here. See my explanation in https://github.com/ethereum/solidity/pull/8164#discussion_r371950090

chriseth commented on tools/yulPhaser/Population.cpp:
```diff
@@ -0,0 +1,140 @@
+		[](auto const& a, auto const& b){ return a.fitness.value() < b.fitness.value(); }
+	);
+
+	randomizeWorstChromosomes(m_individuals, m_individuals.size() / 2);
```
Maybe it is more flexible to do:

```
size_t desiredSize = m_individuals.size();
m_individuals.resize(desiredSize / 2);
refillPool(m_individuals, desiredSize);
```

And then refill does:

```
while (m_individuals.size() < _desiredSize)
  m_individuals.emplace_back(generateNewIndividual(m_individuals));
```

cameel replied:
Yeah. That's a good idea for refactoring the current design but since I'm changing that design in the next PR anyway, I'll leave it as is here.

chriseth commented on tools/yulPhaser/Population.h:
```diff
@@ -0,0 +1,89 @@
+struct Individual
+{
+	Chromosome chromosome;
+	std::optional<size_t> fitness;
```
```suggestion
	std::optional<size_t> fitness = nullopt;
```
to allow Individual being constructed from a Chromosome?

cameel replied:
Done.

chriseth commented on tools/yulPhaser/Program.cpp:
```diff
@@ -0,0 +1,152 @@
+	return _stream << AsmPrinter()(*_program.m_ast);
+}
+
+string Program::toJson() const
```
Do we really need that?

cameel replied:
Not really. But it's very handy for inspecting the Yul code when writing tests or debugging. I could just keep it stashed constantly and apply it when I need it but I thought that it wouldn't hurt to leave it in in case it comes in handy later (it probably will). It's very short so it should not be big maintenance burden.

chriseth replied:
Isn't the AsmPrinter much more useful?

cameel replied:
Depends. Both were useful in different situations. When writing tests I found it very helpful to see the exact AST structure, including the names of the classes used as its nodes. I could figure this out by looking at yul code and then finding the right definitions in `AsmData.h` but it would have taken me more time. And even then I could easily get it wrong - I think it's a good thing to be able to see exactly what's in the tree rather than assume based on my (more or less complete) understanding of it.

chriseth replied:
Ah ok, makes sense given that in your tests you were actually checking the exact structure of the ast instead of just its pretty-printed form.

cameel commented:
> It might be a good idea to extract the changes to yulopti into its own PR.

Sure. I'll put it in a separate one.

> Also could you rebase and squash the rest?

Done.

chriseth commented:
Waiting for you to extract the changes to isoltest to its own PR before reviewing this a final time.

cameel committed:
[yul-phaser] Remove Chromosome::allStepNamesExcept()

cameel commented:
All review fixes done. In places where I think a change is not required or does not make sense because of the design changes in the upcoming PR I left a reply explaining why. I also resolved the comments where the issue was trivial or stuff has been moved to a different PR and will be reviewed there anyway.

 I have extracted the commits changing stuff outside of `yul-phaser` into three small PRs: #8213, #8214, #8215. I see that the first one has already been merged (thanks :)). The PRs are independent but since this one depends on all of them, I put them on top of each other. This unfortunately means that the changes from all of them will show also here. I could normally avoid that by changing the base branch of the PR but it looks like I can't do it if the base branch exists only in a fork.

chriseth commented:
Thanks a lot! Let's now get the small ones merged and then this one so you can finally continue your already ongoing work:)

cameel commented:
Sure. I'm unfortunately starting late today (sorry that I was not available to make changes in #8214) but I'll try to get the remaining two sorted out now.

cameel commented:
I just pushed a few minor cleanup commits extracted from my new branch (as fixups). This is trivial stuff that should not require re-reviewing:
- `Program`'s move constructor made public (was protected by mistake).
- Removed unnecessary `std::` from two `.cpp` files.
- Removed `assert()` that's now always true after a change to `size_t` from `Population::run()`.
- Moved the definition of `Population::measureFitness()` to a better spot in the `.cpp` file.
- Removed a temporary comment left by mistake in `Population::randomizeWorstChromosomes()`.

chriseth commented:
Is this the next PR to review? Can you please rebase and squash?

cameel commented:
Done. All fixups squashed.

Yes, this one is next in line. After it comes a small one - #8223. And after that one a new, bigger pull request with the actual genetic algorithm that I'm aiming to have ready some time tomorrow.

chriseth commented:
Looks good. I'll rebase and then merge.

chriseth commented:
Hm, it seems I still cannot push to the branch. Can you please rebase?

cameel commented:
Sure.

BTW, I found the option to enable it ([Allowing changes to a pull request branch created from a fork](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork)) but it's unavailable on my PRs for some reason and I can't do it here. It's available when I'm submitting a PR from my private fork but not when doing so from `imapp-pl`.

cameel commented:
Rebased on `develop`.

chriseth commented:
There seems to be a memory problem (built with clang):
```
Running /root/project/build/test/soltest --color_output=no --show_progress=yes --logger=JUNIT,error,test_results/constantinople.xml -- --evm-version=constantinople 

0%   10   20   30   40   50   60   70   80   90   100%
|----|----|----|----|----|----|----|----|----|----|
**malloc(): unsorted double linked list corrupted
./.circleci/soltest.sh: line 64:    93 Aborted                 (core dumped) ${REPODIR}/build/test/soltest ${BOOST_TEST_ARGS} -- ${SOLTEST_ARGS}
```

cameel commented:
Ok. I'll check it. All the checks were passing before so it's possible it's from something below. I see a comment saying that the build was failing here: https://github.com/ethereum/solidity/pull/7914#issuecomment-582474231. Has that problem been actually resolved before it was merged?

chriseth commented:
I checked the latest commit in develop and it is fine. I'll retrigger the test.

cameel commented:
Ok. And I'll see what happens when I build my branch locally with clang.

chriseth commented:
Re-run was successful.
